src/v/storage/spill_key_index.cc

```diff
@@ -108,47 +108,54 @@ ss::future<> spill_key_index::add_key(compaction_key b, value_type v) {
     if (
       (take_result.checkpoint_hint && expected_size > min_index_size)
       || expected_size >= _max_mem) {
-        f = ss::do_until(
-          [this, entry_size, min_index_size] {
-              size_t total_mem = idx_mem_usage() + _keys_mem_usage + entry_size;
-
-              // Instance-local capacity check
-              bool local_ok = total_mem < _max_mem;
-
-              // Shard-wide capacity check
-              bool global_ok = _resources.compaction_index_bytes_available()
-                               || total_mem < min_index_size;
-
-              // Stop condition: none of our size thresholds must be violated
-              return _midx.empty() || (local_ok && global_ok);
-          },
-          [this] {
-              /**
-               * Evict first entry, we use hash function that guarante good
-               * randomness so evicting first entry is actually evicting a
-               * pseudo random elemnent
```

Comment:
Well sort of, maybe for 1 element, but it leads to pretty weird biases.

E.g., if you remove half of the elements in the map, then add 1 more, then with 50% chance the just added element will be the next one removed (if it falls in the half of the hash map that has been cleared out), which is a very large bias and does not look at all like random sampling. Not sure if it matters for our usage.

It would be nice if hash maps actually had some support for this, e.g., index into a random bucket and find the closest element or something.